### ISSUE TITLE
After upgrading to 1.10.1 from 1.8 kibana login is not working

### ISSUE THREAD

**Issue by sushovan23** [opendistro-for-elasticsearch/opendistro-build#416]
**Describe the bug**
performed upgrade to 1.10.1 from 1.8. 
All elasticsearch nodes are up and cluster is green. 
Before upgrade cluster was configured with openid/sso. we were able to login correctly.
After upgrade when I try to login to kibana its giving following error:

{"statusCode":400,"error":"Bad Request","message":"[request query.session_state]: definition for this key is missing"}

**To Reproduce**
1. upgrade to 1.10.1 from 1.8  (elasticsearch should be configured
2. upgrade opendistro kibana
3. try to login
{"statusCode":400,"error":"Bad Request","message":"[request query.session_state]: definition for this key is missing"}
**Expected behavior**
Able to login to kibana using sso.

**Configuration (please complete the following information):**
 - ODFE/Kibana version [1.10.1]
 - Distribution [RPM]
 - Host Machine [Centos7]

Did something change in the way security plugin is configured? Because I have it working correctly in 1.6 and 1.8.

**Comment by lukasmrtvy** [comment]
I have same problem, I am using openid_auth_domain and Azure AD

**Comment by lukasmrtvy** [comment]
Related: https://stackoverflow.com/questions/53650185/azure-active-directory-v2-0-doesnt-seem-to-support-oauth-session-management.
Problem ( at least for me and Azure AD ) is that v2 endpoint `https://login.microsoftonline.com/<tenantid>/v2.0/.well-known/openid-configuration` does not support session_state.

**Comment by sushovan23** [comment]
We are using v2.0 in opendistro 1.6 and 1.8. Those are working perfectly. So something must have changed in 1.10 which has caused it to break.

**Comment by gaiksaya** [comment]
Hi @sushovan23 ,

Sorry for the issues faced by upgrading the ODFE to v1.10.1. We have added the instructions to unblock this. Please check https://opendistro.github.io/for-elasticsearch-docs/docs/upgrade/1-10-1/
Please let us know if you continue to face the issue.
Thanks!

**Comment by chandu484** [comment]
Hello @gaiksaya Thanks for your reply!  I'm following this thread as I'm also facing same issue with Okta OIDC connector. Even with that setting its not working! Seeing same error **{“statusCode”:403,“error”:“Forbidden”,“message”:“Request failed authentication.”}**

**Comment by lukasmrtvy** [comment]
@gaiksaya setting `opendistro_security.cookie.secure: false` to `kibana.yml` does not work for me.  Still getting:
```
{
"statusCode": 400,
"error": "Bad Request",
"message": "[request query.session_state]: definition for this key is missing"
}
```

**Comment by gaiksaya** [comment]
Hi @lukasmrtvy ,

Did you restart Kibana after adding that setting?

**Comment by lukasmrtvy** [comment]
@gaiksaya yes, of course. I am using helm chart v1.10.1.

**Comment by peterzhuamazon** [comment]
@vrozov @zengyan-amazon Seems like we have multiple community members with security plugin related issues.
Could you guys help with this?
Thanks.

**Comment by sushovan23** [comment]
Hi @gaiksaya:
I have added setting and restated kibana. still same error

**Comment by zengyan-amazon** [comment]
@sushovan23 can you share your kibana.yml configuration so that we can try to reproduce the issue on our side?

**Comment by sushovan23** [comment]
here is the kibana.yml:
```
# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License is located at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# or in the "license" file accompanying this file. This file is distributed
# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
# express or implied. See the License for the specific language governing
# permissions and limitations under the License.

# Description:
# Default Kibana configuration for Open Distro.

elasticsearch.hosts: https://localhost:9200
elasticsearch.ssl.verificationMode: none
elasticsearch.username: <user>
elasticsearch.password: <pwd>
elasticsearch.requestHeadersWhitelist: ["securitytenant","Authorization", "security_tenant"]

opendistro_security.multitenancy.enabled: true
opendistro_security.multitenancy.tenants.preferred: ["Private", "Global"]
opendistro_security.readonly_mode.roles: ["kibana_read_only"]
opendistro_security.auth.type: "openid"
opendistro_security.openid.connect_url: "https://login.microsoftonline.com/<secret>/v2.0/.well-known/openid-configuration"
opendistro_security.openid.client_id: "<client_id>"
opendistro_security.openid.client_secret: "<secret>"
opendistro_security.openid.base_redirect_url: "https://fqdn/"
opendistro_security.cookie.secure: false
```

these are rpms installed on server:
```
rpm -qa | grep -i opendistr
opendistro-security-1.10.1.0-0.noarch
opendistro-knnlib-1.10.1.0-1.x86_64
opendistroforelasticsearch-kibana-1.10.1-1.x86_64
opendistro-anomaly-detection-1.10.1.0-1.noarch
opendistro-job-scheduler-1.10.1.0-1.noarch
opendistro-performance-analyzer-1.10.1.0-1.noarch
opendistro-alerting-1.10.1.2-1.noarch
opendistro-knn-1.10.1.0-1.noarch
opendistro-sql-1.10.1.1-1.noarch
opendistro-index-management-1.10.1.1-1.noarch
opendistroforelasticsearch-1.10.1-1.noarch
```
elasticsearch version
`elasticsearch-oss-7.9.1-1.x86_64`

Let me know if anything else is needed

**Comment by zengyan-amazon** [comment]
I am suspecting this is due to our new Kibana security plugin OpenIDConnect implementation issue, which doesn't allow `session_state` as query parameter.

can you try updating the `<kibana_installation_dir>/plugins/opendistro_security/server/auth/types/openid/routes.js`, update line 43 - 48 
```
                  query: config_schema_1.schema.object({
                      code: config_schema_1.schema.maybe(config_schema_1.schema.string()),
                      nextUrl: config_schema_1.schema.maybe(config_schema_1.schema.string()),
                      state: config_schema_1.schema.maybe(config_schema_1.schema.string()),
                      refresh: config_schema_1.schema.maybe(config_schema_1.schema.string()),
                  }),
```
to
```
query: config_schema_1.schema.any(),
```
and restart Kibana to see if it works?

**Comment by sushovan23** [comment]
That did it. However first time I logged in after reboot I saw following error:
```
Error: Internal Server Error
    at Fetch._callee3$ (https://fqdn/33912/bundles/core/core.entry.js:34:109213)
    at l (https://fqdn/33912/bundles/kbn-ui-shared-deps/kbn-ui-shared-deps.js:368:155323)
    at Generator._invoke (https://fqdn/33912/bundles/kbn-ui-shared-deps/kbn-ui-shared-deps.js:368:155076)
    at Generator.forEach.e.<computed> [as next] (https://fqdn/33912/bundles/kbn-ui-shared-deps/kbn-ui-shared-deps.js:368:155680)
    at fetch_asyncGeneratorStep (https://fqdn/33912/bundles/core/core.entry.js:34:102354)
    at _next (https://fqdn/33912/bundles/core/core.entry.js:34:102670)
```

Also noticed that Global tenant was not available for non admin users? Is that a bug & should be reported separately?

**Comment by sushovan23** [comment]
![image](https://user-images.githubusercontent.com/14843483/95138093-6b510680-0737-11eb-9d10-da0c9e99dd95.png)

**Comment by sushovan23** [comment]
As you can see Global tenant is grayed out.

**Comment by zengyan-amazon** [comment]
The error message seems not related to the OpenIDConnect issue, we will also try to reproduce it on our side.

And I would suggest to open a separate issue for the global tenant issue. please create the issue in Kibana security plugin repo https://github.com/opendistro-for-elasticsearch/security-kibana-plugin/issues

**Comment by sushovan23** [comment]
yes I only noticed it first time i logged in. Havent seen it since.

I have opened separate issue for global tenant grayed out:
https://github.com/opendistro-for-elasticsearch/security-kibana-plugin/issues/509

**Comment by zengyan-amazon** [comment]
@chandu484 can you try clear your browser cookies, or try logging in using Incognito mode? if it still not working, please share your `kibana.yml` to help us diagnose. Thanks!

**Comment by sushovan23** [comment]
Will we get patch released or minor release with this fix as we want to use 1.10.x?

**Comment by zengyan-amazon** [comment]
@sushovan23 We are working on creating a 1.10.1.2 release of security kibana plugin, so that you can download and replace the plugin. Hope this works for you.

**Comment by sushovan23** [comment]
Great we are presently installing opendistroforelasticsearch-kibana rpm and I mention the version number 1.10.1 so I’m hoping all I need to do is to update version 1.10.1.2 for that rpm and elastic search related remains same at 1.10.1 or will that change. I’ll wait for release n install/upgrade documentation to figure it out. Thank you

**Comment by chandu484** [comment]
> @chandu484 can you try clear your browser cookies, or try logging in using Incognito mode? if it still not working, please share your `kibana.yml` to help us diagnose. Thanks!

Hi @zengyan-amazon I have followed the steps you mentioned to modify routes.js in kibana node and after that I'm not seeing any issue with iss in kibana. Looks good. Thanks for that. I'm using amazon/opendistro kibana image 1.10.1. How can we update the patch for kibana plugin (1.10.1.2) you are taking about in the container as its using base image?

**Comment by lukasmrtvy** [comment]
Is there any private version of v1.10.1.2 kibana image to test ? Thanks

**Comment by arnitolog** [comment]
for some reason code substitution didn't help me. I run everything on docker, so I deleted the original plugin first, and then install the new one, which has the above changes. And I got an error in the browser: "ERR_TOO_MANY_REDIRECTS"

**Comment by zengyan-amazon** [comment]
@arnitolog can you clear your browser cookies or try in incognito mode?

**Comment by arnitolog** [comment]
tried in incognito and even in a different browser. But got the same result - the error above.

**Comment by zengyan-amazon** [comment]
@arnitolog can you share your configurations in kibana.yml and your ES security OIDC config? it will help us to understand your setup

**Comment by arnitolog** [comment]
sure. Here is kibana.yml:

`
   
    server.name: kibana
    server.host: "0"
    ## Replace with Elasticsearch DNS name picked during Service deployment
    elasticsearch.hosts: https://es.local:9200

    ## Kibana TLS Config
    # server.ssl.enabled: true
    # server.ssl.key: /usr/share/kibana/certs/kibana-key.pem
    # server.ssl.certificate: /usr/share/kibana/certs/kibana-crt.pem
    elasticsearch.ssl.verificationMode: none
    elasticsearch.ssl.certificate: /new/certificates/tls.crt
    elasticsearch.ssl.key: /new/certificates/tls.key
    elasticsearch.ssl.certificateAuthorities: /new/certificates/ca.crt
    elasticsearch.username: kibanaserver
    elasticsearch.password: xxxxxxxxxxxx
    elasticsearch.requestTimeout: 150000
    
    logging.dest: stdout
    logging.json: true
    logging.timezone: UTC
    logging.verbose: false

    elasticsearch.requestHeadersWhitelist: ["securitytenant","Authorization"]
    opendistro_security.allow_client_certificates: true
    opendistro_security.multitenancy.enabled: false
    opendistro_security.multitenancy.tenants.preferred: ["Global","Private"]
    opendistro_security.readonly_mode.roles: ["kibana_read_only"]
    opendistro_security.auth.type: "openid"
    opendistro_security.openid.connect_url: "https://keycloak.company.com/auth/realms/master/.well-known/openid-configuration"
    opendistro_security.openid.base_redirect_url: "https://logs.dev.company.com"
    opendistro_security.openid.client_id: "kibana-dev"
    opendistro_security.openid.client_secret: "zzzzzzzzzzzzzzzz"
    opendistro_security.cookie.password: XXXXXXXXXXXXXXX
    opendistro_security.cookie.ttl: 3600000
    opendistro_security.session.ttl: 86400000
    opendistro_security.session.keepalive: true
    opendistro_security.cookie.secure: false`

Here is security config:

```config:
       dynamic:
        http:
          anonymous_auth_enabled: false
          xff:
            enabled: false
        authc:
          basic_internal_auth_domain:
            http_enabled: true
            order: 0
            http_authenticator:
              type: basic
              challenge: false
            authentication_backend:
              type: internal
          openid_auth_domain:
            http_enabled: true
            order: 1
            http_authenticator:
              type: openid
              challenge: true
              config:
                subject_key: preferred_username
                roles_key: roles
                openid_connect_url: https://keycloak.company.com/auth/realms/master/.well-known/openid-configuration
            authentication_backend:
              type: noop

**Comment by plnordquist** [comment]
I had the `{"statusCode":400,"error":"Bad Request","message":"[request query.session_state]: definition for this key is missing"}` error when I connected my Kibana to Keycloak and was able to work around it by configuring the client registration in Keycloak to turn on the Exclude Session State From Authentication Response option.  Is there a point where I will need to reverse this configuration?  Keycloak alludes to this option being useful for maintaining compatibility with older ODIC/OAuth 2 adapters.

![session-state](https://user-images.githubusercontent.com/8883846/95246806-e19e4900-07c9-11eb-89f2-9be97be3bfac.png)
